### PR TITLE
add abilitiy to destroy guard conditions and destroy the sigint guard condition in spin_once

### PR DESCRIPTION
**Pull request by mikaelarguedas**
This fixes partially #74

### CODE REVIEW

File: rclpy/src/rclpy/_rclpy.c [hunk 1]
```diff
@@ -1434,6 +1434,10 @@ rclpy_destroy_entity(PyObject * Py_UNUSED(self), PyObject * args)
   } else if (0 == strcmp(entity_type, "timer")) {
     rcl_timer_t * timer = (rcl_timer_t *)p;
     ret = rcl_timer_fini(timer);
+  } else if (0 == strcmp(entity_type, "guard_condition")) {
+    rcl_guard_condition_t * guard_condition =
+      (rcl_guard_condition_t *)PyCapsule_GetPointer(pyentity, NULL);
```
Comment (dirk-thomas) [inline rclpy/src/rclpy/_rclpy.c]: `PyCapsule_GetPointer(pyentity, NULL)` should be replaced by `p`.

Then this will also fit in one line.

Comment (mikaelarguedas) [reply rclpy/src/rclpy/_rclpy.c]: oh I copy pasted before rebasing :S

**Comment by dirk-thomas** [comment]
The remaining memory usage is still related to the `rclpy_get_sigint_guard_condition` call. So this needs some more changes.

File: rclpy/rclpy/client.py [hunk 2]
```diff
@@ -37,6 +37,9 @@ def run(self):
 
         guard_condition_ready_list = \
             _rclpy.rclpy_get_ready_entities('guard_condition', self.wait_set)
+
+        # destroying here to make sure we dont call shutdown before cleaning up
+        _rclpy.rclpy_destroy_entity('guard_condition', sigint_gc)
```
Comment (dirk-thomas) [inline rclpy/rclpy/client.py]: Won't this invalidate the variable `sigint_gc_handle` which is used next line?

Comment (mikaelarguedas) [reply rclpy/rclpy/client.py]: the pointer / handle is stored in a python int and is compared to a list of already populated int so I don't think that's a problem. But maybe I'm overseeing this

Comment (dirk-thomas) [reply rclpy/rclpy/client.py]: I see, while the pointer is invalid the integer value is still the same and can be used for the following comparison. :+1:

**Comment by dirk-thomas** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/3168/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/508/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2525/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/3222/)

**Comment by leander2189** [comment]
Can this merge be backported to humble?

**Comment by fujitatomoya** [comment]
@leander2189 

this PR was merged in 2027, that is a way before humble is introduced. that said, all current distros should have this fix. did you actually find the same problem on humble?

**Comment by leander2189** [comment]
Sorry @fujitatomoya this comment was meant for MR #1308, somehow I ended up here.

Again, sorry for the disturbance.

**Pull request merged**